File: order-api/services/orderService.js
const orderRepo = require('../repositories/orderRepository');
const eventBus = require('../events/eventBus');

class OrderService {
  async createOrder(data) {
    const order = await orderRepo.create(data);

    await eventBus.publish('OrderPlaced', {
      orderId: order._id,
      userId: order.userId,
      items: order.items.map(item => ({
        productId: item.productId,
        quantity: item.quantity
      })),
      createdAt: order.createdAt
    });

    return order;
  }

  getAllOrders() {
    return orderRepo.getAll();
  }

  getOrderById(id) {
    return orderRepo.getById(id);
  }

  getOrdersByUserId(userId) {
    return orderRepo.getByUserId(userId);
  }
}

module.exports = new OrderService();

File: order-api-net/Order.Api/Services/OrderService.cs
public class OrderService
{
    private readonly OrderRepository _repo;
    private readonly EventBus _eventBus;

    public OrderService(OrderRepository repo, EventBus eventBus)
    {
        _repo = repo;
        _eventBus = eventBus;
    }
    
    public async Task<Order> CreateOrder(Order order)
    {
        await _repo.Create(order);

        _eventBus.Publish(new
        {
            orderId = order.Id,
            userId = order.UserId,
            items = order.Items.Select(i => new { i.ProductId, i.Quantity }),
            createdAt = order.CreatedAt
        });

        return order;
    }

    public Task<List<Order>> GetAllOrders() => _repo.GetAll();
    public Task<Order?> GetOrderById(string id) => _repo.GetById(id);
    public Task<List<Order>> GetOrdersByUserId(string userId) => _repo.GetByUserId(userId);
}
